src/core/Footer/Footer.tsx

```diff
@@ -24,6 +24,10 @@ export type FooterProps = {
    * Provide localized strings.
    */
   translatedTitles?: TitleTranslations;
+  /**
+   * If true, only custom elements will be shown.
+   */
+  showOnlyCustomElements?: boolean;
```

Comment:
So I was thinking about this prop and it bothers me, cause its really specific and not extendable..
I was talking with @mayank99 and he suggested something like that
```js
customElements?: FooterElement[] | (defaultElements: FooterElement[]) => FooterElement[];
```
So basically, we would accept now prop `customElements` as function, where we would pass our default ones (for first implementation might not) and expect user to give us a list of `FooterElement` which we would render straight away as is (without any checks). That way user could pass the list of needed elements and we would not include the default ones.
The only issue for `defaultElements` as function prop is the Bentley copyright part which now is toggled by the new flag. Then it would be removed, I guess?

Do you think you could try to play around a little with the API as see what you could do without a new prop, but instead extending the current API?

Reply:
Maybe we should include the Bentley copyright part in `defaultElements`

Reply:
My thoughts are:

1. This is potentially more confusing and difficult to use and needs more explanation than a true/false prop.
2. Another way to implement this might be to have an `options` prop that includes `showOnlyCustomElements` and could include more options in the future, which would make it less specific and more extendable.
3. I understand that doing it your suggested way would allow users more control over which default elements are in the footer, and in what order. However, I think that there are two scenarios for the Footer: it is being used in a Bentley product or it's not. In the case of it being used in a Bentley product, the default elements are required and I think it's better for the order and placement of the default elements to be the same across all Bentley products. In the case of the Footer not being in a Bentley product, users won't need any of the default elements. So while this idea provides more flexibility, I think that flexibility is uneccessary and could potentially lead to less cohesion across Bentley products.

But I'm new to this team, so if you think that your way is better, I'll do it that way.

Reply:
1. I do understand this is not something used widely, but actually it is not something fancy. Users either pass array or they get to implement function, where they could fully customize footer content whereas we ensure the correct styling. Explanation is needed anyway, so be sure to have a good prop description and story example will be helpful too.
Whenever you use a component, you will need to get familiar with its API and to achieve some custom/advanced case it might always need not regular implementation.
2. This is something better and worth thinking.. Just the issue with it is the usage, which is not nice. As the prop gonna be object, if user wants to change something, he will need to create a new object reference everytime to make footer react to it. If you want, try to apply this suggestion and see, how it looks. We could talk about it more.
3. That is a really logical thinking, but few points to mention from experience.
- In our default links there are some, which need authentication to be opened (check the first one, for example). We might have products, which do not want to have such legal links, cause users can fully use the product without authentication (developer portal as an example). So such products might want to customize the links, by changing URL, for example
- Bentley also has a lot of acquired products, which not necessarily are fully under Bentley and considered as Bentley product. They might still work under their own environment including legal rights and we could expect them having different links to certain aspects of legal
- And as usual, there might be some weird and not thought through use cases, where usually we just try to have as customizable solution as possible, just to be safe, because we already had some decisions made which now we already need to change. You can check our [v2 wiki page](https://github.com/iTwin/iTwinUI-react/wiki/v2-release), to see the collection of needed breaking changes.

Reply:
I am writing a story for a use case where default and custom elements are being intermixed. In doing so I realized that a problem with this implementation is that it's going to be difficult for users to know what is incoming in the `defaultElements` parameter.

I initially used the array position, but for this to be intuitive for the users there will need to be documentation that is kept up to date. I know that we have documentation; does it include information to this level of detail? I feel like outlining what is in each position is too much information to be added to the prop description.

I also experimented with using `.find()` on the `defaultElements` parameter to get the element by the `title` property. For this to be simple, users will have to use a non-null assertion. Otherwise, they'll have to null check every time they use a value from `defaultElements`. There would also need to be documentation for the titles of each default element, although that might be succinct enuogh to include in the prop description.

Do you have an alternative method of using the `defaultElements` parameter?

Reply:
Honestly, I don't find the below to be *too* difficult or complicated or unreadable. Array operations are pretty common.

Only custom:
```tsx
const newFooterParts = [ /* populate this with new stuff */ ];
<Footer customElements={() => newFooterParts} />
```

Appending custom unchanged:
```tsx
const newFooterParts = [ /* populate this with new stuff */ ];
<Footer customElements={(defaultParts) => [...defaultParts, ...newFooterParts]} />
```

Filtering out:
```tsx
const translatedTiles = { /* localized */ };
<Footer
  translatedTitles={translatedTitles}
  customElements={(defaultParts) => defaultParts.filter(({ title }) => title !== translatedTitles.termsOfService)}
/>
```

Changing one url:
```tsx
const translatedTitles = { /* localized */ };
const customPrivacyUrl = '...';
<Footer
  translatedTitles={translatedTitles}
  customElements={(defaultParts) => defaultParts.map(d => ({ ...d, url: d.title === translatedTitles.privacy ? customPrivacyUrl : d.url }))}
/>
```

Changing all urls:
```tsx
const translatedTitles = { /* localized */ };
const customUrls = { [translatedTitles.privacy]: '...', /* other urls using values of titles as keys */  };
<Footer
  translatedTitles={translatedTitles}
  customElements={(defaultParts) => defaultParts.map(({ title, url }) => ({ title: title, url: customUrls[title] ?? url }))}
/>
```

However, I agree that using `title` is weird so maybe we could expose a `key` in each of our defaultElements. It would simplify the last example like this (notice how we don't even need the `translatedTitles` prop anymore):
```tsx
const translatedTitles = { /* localized */ };
const customUrls = { privacy: '...', /* urls using actual keys */  };
<Footer
  customElements={(defaultParts) => defaultParts.map(({ key }) => ({ key, title: translatedTitles[key], url: customUrls[key] }))}
/>
```

Reply:
This is a very nice list of examples!
need to store it somewhere :D
Title looks weird to me too and story has null assertions all over the place now.

Reply:
These are pretty good solutions for most cases, and the key is an excellent idea. However, this doesn't solve the difficulty of reordering elements. I suppose that's not a big concern though.

Reply:
But `key` type now is weird.. What if user passes custom elements? key will not be from translations if he wants to add it..
Then I guess something like
```js
key?: keyof TitleTranslations | string;

Reply:
I thought about this, but thought that the (intended) use for keys is in accessing the default elements. I can imagine that someone might want to utilize them though.

Reply:
A negative to adding `| string` to the key definition is that it removes the ability to see what the defined keys, and thus the default elements, are. Which is important for ease of use for users.

Is it really terrible for it to stay as is?

Reply:
well, we export this type for users to use for custom element.
And we do not allow to set any other key to them, so if they have custom element, they cannot set key (cant use the default one).
As I mentioned, either allow to pass in key anything or change the type for external use without key.

Do you have other suggestions?

Reply:
> A negative to adding `| string` to the key definition is that it removes the ability to see what the defined keys, and thus the default elements, are.

This is a bug in typescript, can be worked around by adding `| {}` or `| Record<never, never>`. See example in https://github.com/iTwin/iTwinUI-react/blob/main/src/core/Badge/Badge.tsx#L53